zcash_client_sqlite/src/lib.rs

```diff
@@ -387,62 +386,108 @@ impl<'a, P: consensus::Parameters> WalletWrite for DataConnStmtCache<'a, P> {
         }
     }
 
-    fn insert_block(
+    fn insert_pruned_block(
         &mut self,
-        block_height: BlockHeight,
-        block_hash: BlockHash,
-        block_time: u32,
-        commitment_tree: &CommitmentTree<Node>,
-    ) -> Result<(), Self::Error> {
-        wallet::insert_block(self, block_height, block_hash, block_time, commitment_tree)
-    }
+        block: &PrunedBlock,
+        updated_witnesses: &[(Self::NoteRef, IncrementalWitness<Node>)],
+    ) -> Result<Vec<(Self::NoteRef, IncrementalWitness<Node>)>, Self::Error> {
+        // database updates for each block are transactional
+        self.transactionally(|up| {
+            // Insert the block into the database.
+            wallet::insert_block(
+                up,
+                block.block_height,
+                block.block_hash,
+                block.block_time,
+                &block.commitment_tree,
+            )?;
+
+            let mut new_witnesses = vec![];
+            for tx in block.transactions {
+                let tx_row = wallet::put_tx_meta(up, &tx, block.block_height)?;
+
+                // Mark notes as spent and remove them from the scanning cache
+                for spend in &tx.shielded_spends {
+                    wallet::mark_spent(up, tx_row, &spend.nf)?;
+                }
 
-    fn rewind_to_height(&mut self, block_height: BlockHeight) -> Result<(), Self::Error> {
-        wallet::rewind_to_height(self.wallet_db, block_height)
-    }
+                for output in &tx.shielded_outputs {
+                    let received_note_id = wallet::put_received_note(up, output, tx_row)?;
 
-    fn put_tx_meta(
-        &mut self,
-        tx: &WalletTx,
-        height: BlockHeight,
-    ) -> Result<Self::TxRef, Self::Error> {
-        wallet::put_tx_meta(self, tx, height)
+                    // Save witness for note.
+                    new_witnesses.push((received_note_id, output.witness.clone()));
+                }
+            }
+
+            // Insert current new_witnesses into the database.
+            for (received_note_id, witness) in updated_witnesses.iter().chain(new_witnesses.iter())
+            {
+                if let NoteId::ReceivedNoteId(rnid) = *received_note_id {
+                    wallet::insert_witness(up, rnid, witness, block.block_height)?;
+                }
```

Comment:
This doesn't return an error like the old code did:
```suggestion
                if let NoteId::ReceivedNoteId(rnid) = *received_note_id {
                    wallet::insert_witness(up, rnid, witness, block.block_height)?;
                } else {
                    Err(SqliteClientError::InvalidNoteId)
                }
```